server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,449 @@
+    private final MasterHistoryService masterHistoryService;
+    /**
+     * This is the amount of time we use to make the initial decision -- have we seen a master node in the very recent past?
+     */
+    private final TimeValue veryRecentPast;
+    /**
+     * If the master transitions from a non-null master to a null master at least this many times it starts impacting the health status.
+     */
+    private final int unacceptableNullTransitions;
+    /**
+     * If the master transitions from one non-null master to a different non-null master at least this many times it starts impacting the
+     * health status.
+     */
+    private final int unacceptableIdentityChanges;
+
+    private static final Logger logger = LogManager.getLogger(StableMasterHealthIndicatorService.class);
+
+    // This is the default amount of time we look back to see if we have had a master at all, before moving on with other checks
+    private static final TimeValue DEFAULT_VERY_RECENT_PAST = new TimeValue(30, TimeUnit.SECONDS);
+    private static final TimeValue SMALLEST_ALLOWED_VERY_RECENT_PAST = new TimeValue(1, TimeUnit.SECONDS);
+
+    // This is the default number of times that it is OK to have a master go null. Any more than this will be reported as a problem
+    private static final int DEFAULT_ACCEPTABLE_NULL_TRANSITIONS = 3;
+    private static final int SMALLEST_ALLOWED_ACCEPTABLE_NULL_TRANSITIONS = 0;
+
+    // This is the default number of times that it is OK to have a master change identity. Any more than this will be reported as a problem
+    private static final int DEFAULT_ACCEPTABLE_IDENTITY_CHANGES = 3;
+    private static final int SMALLEST_ALLOWED_ACCEPTABLE_IDENTITY_CHANGES = 0;
+
+    // Keys for the details map:
+    private static final String DETAILS_CURRENT_MASTER = "current_master";
+    private static final String DETAILS_RECENT_MASTERS = "recent_masters";
+    private static final String DETAILS_EXCEPTION_FETCHING_HISTORY = "exception_fetching_history";
+
+    // Impacts of having an unstable master:
+    private static final String UNSTABLE_MASTER_INGEST_IMPACT = "The cluster cannot create, delete, or rebalance indices, and cannot "
+        + "insert or update documents.";
+    private static final String UNSTABLE_MASTER_DEPLOYMENT_MANAGEMENT_IMPACT = "Scheduled tasks such as Watcher, ILM, and SLM will not "
+        + "work. The _cat APIs will not work.";
+    private static final String UNSTABLE_MASTER_BACKUP_IMPACT = "Snapshot and restore will not work.";
+
+    public static final Setting<TimeValue> VERY_RECENT_PAST_SETTING = Setting.timeSetting(
+        "health.master_history.very_recent_past",
+        DEFAULT_VERY_RECENT_PAST,
+        SMALLEST_ALLOWED_VERY_RECENT_PAST,
+        Setting.Property.NodeScope
+    );
+
+    public static final Setting<Integer> ACCEPTABLE_NULL_TRANSITIONS_SETTING = Setting.intSetting(
+        "health.master_history.acceptable_null_transitions",
```

Comment:
I don't think we should "leak" the word/concept of 'null' here, perhaps "`acceptable_no_master_transitions`" or "`acceptable_none_transitions`"?

Reply:
I changed the setting name and the setting variable name, but left internal variable names and method names as they were b/c using the word "null" really makes it easier to follow internally. Let me know if you disagree.